### PR TITLE
Remote Environments: Integrates remote environment into `azd` commands

### PR DESCRIPTION
**Pull request by wbreza**
Integrates new remote environment components into `azd` commands & components.

## Integration Concepts

- Refactor the environment loading and saving to `environment.Manager`
- Migration from `environment.Ephemeral()` to `environment.New()` or `environment.NewWithValues()`
- Injection of `environment.Manager` into various structs

## How does remote state work?

When remote state is enabled environment state is automatically persisted to the configured remote store.  Any `azd` command that writes to `azd` environment or `azd` environment configuration is automatically persisted.

### Azure Blob Storage

`azd` writes `.env` and `config.json` files to an Azure storage blob container

#### Configuration
- **accountName**: Name of the Azure storage account
- **containerName**: Name of the container within the storage account where configuration is stored. Defaults to the current azd project name if not specified
- **endpoint**: Azure Endpoint used when configuring remote state. _Defaults to `blob.core.windows.net`_

## Configuring remote state

Remote state for `azd` can either be configured globally in azd's `config.json` or per project within the `azure.yaml`.  If remote state is not configured, environment values and configuration are still stored locally.

## Schema

Remote state is configured within the `state.remote` element of `azd` configuration

- **backend**: The name of the backend type used for remote state
- **config**: Map of key/value pairs unique to each remote state provider

## azure.yaml

```yaml
name: azd-project-name
state:
  remote:
    backend: AzureStorage
    config:
      accountName: saazdremotestate
      containerName: myproject # Defaults to project name if not specified
```

### azd config.json

```json
{
  "state": {
    "remote": {
      "backend": "AzureStorage",
      "config": {
        "accountName": "saazdremotestate"
      }
    }
  }
}
```

### `azd env list`

Lists all local and remote environments available

image

### `azd env select`

When selecting an environment that does not exist locally, the remote state is copied to a new local environment.

### CODE REVIEW

File: cli/azd/cmd/util.go [hunk 1]
```diff
@@ -87,147 +86,6 @@ type environmentSpec struct {
 	examples []string
 }
 
-// createEnvironment creates a new named environment. If an environment with this name already
-// exists, and error is returned.
-func createEnvironment(
```
Comment (wbreza) [inline cli/azd/cmd/util.go:92]: Moved into environment Manager.

File: cli/azd/cmd/util_test.go [hunk 2]
```diff
@@ -2,105 +2,25 @@ package cmd
...
-			azdContext,
-			mockContext.Console,
-		)
-		require.ErrorContains(
-			t,
-			err,
-			fmt.Sprintf("environment '%s' already exists",
-				validName))
-	})
-}
```
Comment (wbreza) [inline cli/azd/cmd/util_test.go:97]: Moved into environment Manager.

File: cli/azd/pkg/environment/environment.go [hunk 3]
```diff
@@ -134,65 +121,6 @@ func CleanName(name string) string {
...
-	env := Ephemeral()
-
-	if values != nil {
-		env.dotenv = values
-	}
-
-	env.dotenv[EnvNameEnvVarName] = name
-
-	return env
-}
```
Comment (wbreza) [inline cli/azd/pkg/environment/environment.go:194]: `environment` objects are always ephemeral now and the loading/saving aspects have been moved to the environment `DataStore` implementations.

**Review by savannahostrowski — APPROVED** [review]

File: cli/azd/pkg/environment/manager.go [hunk 4]
```diff
@@ -317,6 +320,15 @@ func (m *manager) Reload(ctx context.Context, env *Environment) error {
 // ensureValidEnvironmentName ensures the environment name is valid, if it is not, an error is printed
 // and the user is prompted for a new name.
 func (m *manager) ensureValidEnvironmentName(ctx context.Context, spec *Spec) error {
+	exampleText := ""
```
Comment (ellismg) [inline cli/azd/pkg/environment/manager.go:323]: Likely doesn't matter in this case, because the number of examples is small, but [`strings.Builder`](https://pkg.go.dev/strings#Builder) can also be useful in cases like this since it is "mutable".

File: cli/azd/pkg/environment/manager.go [hunk 5]
```diff
@@ -325,8 +337,7 @@ func (m *manager) ensureValidEnvironmentName(ctx context.Context, spec *Spec) er
 
 			This value is typically used by the infrastructure as code templates to name the resource group that contains
 			the infrastructure for your application and to generate a unique suffix that is applied to resources to prevent
-			naming collisions.`),
-			DefaultValue: spec.Suggest,
```
Comment (ellismg) [inline cli/azd/pkg/environment/manager.go:329]: Just checking - was it intentional that we removed `DefaultValue: spec.Suggest` here?

File: cli/azd/pkg/infra/provisioning/bicep/prompt.go [hunk 6]
```diff
@@ -4,13 +4,13 @@ import (
...
 	"encoding/json"
 	"fmt"
-	"slices"
 	"strconv"
 
 	"github.com/azure/azure-dev/cli/azd/pkg/account"
 	"github.com/azure/azure-dev/cli/azd/pkg/azure"
 	"github.com/azure/azure-dev/cli/azd/pkg/input"
 	"github.com/azure/azure-dev/cli/azd/pkg/output"
+	"golang.org/x/exp/slices"
```
Comment (ellismg) [inline cli/azd/pkg/infra/provisioning/bicep/prompt.go:13]: NIT: I think you can undo this change.

**Review by ellismg — APPROVED** [review]
A few small nits but nothing blocking.  Great work here, @wbreza!

File: cli/azd/pkg/environment/manager.go (hunk 5 above)
Comment (wbreza) [reply cli/azd/pkg/environment/manager.go:329]: Likely a rebase issue if not intended since there was some refactoring in this area.
